nano/node/monitor.cpp

```diff
@@ -0,0 +1,126 @@
+#include "nano/secure/ledger.hpp"
+
+#include <nano/lib/thread_roles.hpp>
+#include <nano/lib/utility.hpp>
+#include <nano/node/monitor.hpp>
+#include <nano/node/node.hpp>
+
+nano::monitor::monitor (nano::monitor_config const & config_a, nano::node & node_a) :
+	config{ config_a },
+	node{ node_a },
+	logger{ node_a.logger }
+{
+}
+
+nano::monitor::~monitor ()
+{
+	debug_assert (!thread.joinable ());
+}
+
+void nano::monitor::start ()
+{
+	if (!config.enabled)
+	{
+		return;
+	}
+
+	thread = std::thread ([this] () {
+		nano::thread_role::set (nano::thread_role::name::monitor);
+		run ();
+	});
+}
+
+void nano::monitor::stop ()
+{
+	{
+		nano::lock_guard<nano::mutex> guard{ mutex };
+		stopped = true;
+	}
+	condition.notify_all ();
+	if (thread.joinable ())
+	{
+		thread.join ();
+	}
+}
+
+void nano::monitor::run ()
+{
+	std::unique_lock<nano::mutex> lock{ mutex };
+	while (!stopped)
+	{
+		run_one ();
+		condition.wait_until (lock, std::chrono::steady_clock::now () + config.interval, [this] { return stopped; });
+	}
+}
+
+void nano::monitor::run_one ()
+{
+	// Node status:
+	// - blocks (confirmed, total)
+	// - blocks rate (over last 5m, peak over last 5m)
+	// - peers
+	// - stake (online, peered, trended, quorum needed)
+	// - elections active (normal, hinted, optimistic)
+	// - election stats over last 5m (confirmed, dropped)
+
+	auto const now = std::chrono::steady_clock::now ();
+	auto blocks_cemented = node.ledger.cemented_count ();
+	auto blocks_total = node.ledger.block_count ();
+
+	// Wait for node to warm up before logging
+	if (last_time != std::chrono::steady_clock::time_point{})
```

Comment:
last_time could be initialized to ::now() which would have the same effect without an if statement.

Reply:
Wouldn't we need an if statement that checks if elasped is not equal 0? Also, delta between last blocks would be incorrect in that case.